src/dwt.rs

```diff
@@ -0,0 +1,184 @@
+//! Debug and trace and stuff
+
+use crate::rcc::Clocks;
+use crate::time::Hertz;
+use cortex_m::peripheral::{DCB, DWT};
+use embedded_hal::blocking::delay::{DelayMs, DelayUs};
+
+pub trait DwtExt {
+    fn constrain(self, dcb: DCB, clocks: Clocks) -> Dwt;
+}
+impl DwtExt for DWT {
+    /// Setup must be called at least once before using other dwt functionality
+    fn constrain(mut self, mut dcb: DCB, clocks: Clocks) -> Dwt {
+        cortex_m::interrupt::free(|_| {
+            // Enable fmc clock
+            dcb.enable_trace();
+
+            self.enable_cycle_counter();
+        });
+
+        Dwt {
+            dwt: self,
+            dcb,
+            clocks,
+        }
+    }
+}
+
+pub struct Dwt {
+    dwt: DWT,
+    dcb: DCB,
+    clocks: Clocks,
+}
+impl Dwt {
+    pub fn release(self) -> (DWT, DCB) {
+        (self.dwt, self.dcb)
+    }
+    pub fn delay(&self) -> Delay {
+        Delay {
+            clock: self.clocks.hclk(),
+        }
+    }
+    pub fn stopwatch<'i>(&self, times: &'i mut [u32]) -> StopWatch<'i> {
+        StopWatch::new(times, self.clocks.hclk())
+    }
+    pub fn measure<F: FnOnce()>(&self, f: F) -> ClockDuration {
+        let mut times: [u32; 2] = [0; 2];
+        let mut sw = self.stopwatch(&mut times);
+        f();
+        sw.lap().lap_time(1).unwrap()
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Delay {
+    clock: Hertz,
+}
+impl Delay {
+    pub fn delay(duration: ClockDuration) {
+        let cycles = duration.ticks as u64;
+        Delay::delay_(DWT::get_cycle_count(), cycles);
+    }
+    fn delay_us_(self, us: u64) {
+        let start = DWT::get_cycle_count();
+        let cycles = (us * self.clock.0 as u64) / 1_000_000;
+        Delay::delay_(start, cycles);
+    }
+    fn delay_ms_(self, ms: u64) {
```

Comment:
Why the trailing underscores? Can't you make those be just the regular `delay_ms` and `pub` and simply remove the macro specialisation for `u64` further down?